File: messenger/message.py
import base64
import struct
from collections import namedtuple

# External encryption methods (you already have these in your code)
from messenger.aes import decrypt, encrypt

# ---------------------------
# 1. Named Tuple Definitions
# ---------------------------

CheckInMessage = namedtuple('CheckInMessage', ['messenger_id'])
InitiateTCPClientReq = namedtuple('InitiateTCPClientReq', ['client_id', 'destination_host', 'destination_port', 'listening_host', 'listening_port'], defaults=['', 0])
InitiateTCPClientRep = namedtuple('InitiateTCPClientRep', ['client_id', 'bind_address', 'bind_port', 'address_type', 'reason', 'remote_addr', 'remote_port'], defaults=['', 0])
SendDataMessage = namedtuple('SendDataMessage', ['client_id', 'data'])
InitiateBINDReq = namedtuple('InitiateBINDReq', ['bind_id', 'listening_host', 'listening_port', 'destination_host', 'destination_port'])
InitiateBINDRep = namedtuple('InitiateBINDRep', ['bind_id', 'listening_host', 'listening_port', 'reason'])
CheckOutMessage = namedtuple('CheckOutMessage', [])

# You could also store message_type inside each namedtuple, or convert them to @dataclass if you prefer.


# --------------------------------
# 2. MessageParser: Reading Bytes
# --------------------------------

class MessageParser:
    @staticmethod
    def read_uint32(data: bytes) -> (int, bytes):
        """
        Reads the first 4 bytes as an unsigned 32-bit integer (big-endian),
        returns (the_integer, remaining_bytes).
        """
        unsigned_32bit = data[:4]               # The 4-byte integer
        remaining_data = data[4:]               # Everything after the 4 bytes
        (value,) = struct.unpack('!I', unsigned_32bit)
        return value, remaining_data

    @staticmethod
    def read_string(data: bytes) -> (str, bytes):
        """
        Reads a length-prefixed UTF-8 string from data:
          1) read an unsigned 32-bit length
          2) read 'length' bytes as the string
        returns (string, remaining_bytes).
        """
        length, data = MessageParser.read_uint32(data)
        s = data[:length].decode('utf-8')
        return s, data[length:]

    @staticmethod
    def parse_check_in(value: bytes) -> CheckInMessage:
        """
        Given decrypted bytes for a 0x04 message,
        read the messenger_id string into a CheckInMessage.
        """
        messenger_id, _ = MessageParser.read_string(value)
        return CheckInMessage(messenger_id=messenger_id)

    @staticmethod
    def parse_initiate_tcp_client_req(value: bytes) -> InitiateTCPClientReq:
        """
        For message type 0x01, parse out:
          - client_id (str)
          - destination_host (str)
          - destination_port (uint32)
        """
        client_id, value = MessageParser.read_string(value)
        destination_host, value = MessageParser.read_string(value)
        destination_port, value = MessageParser.read_uint32(value)
        # listening_host / listening_port are optional — appended by a remote
        # port forwarder so the server can map the forwarded connection to the
        # exact RPF by its listening endpoint. Absent for server-initiated
        # (SOCKS/local) requests.
        listening_host = ''
        listening_port = 0
        if len(value) > 0:
            listening_host, value = MessageParser.read_string(value)
            listening_port, value = MessageParser.read_uint32(value)
        return InitiateTCPClientReq(
            client_id=client_id,
            destination_host=destination_host,
            destination_port=destination_port,
            listening_host=listening_host,
            listening_port=listening_port
        )

    @staticmethod
    def parse_initiate_tcp_client_rep(value: bytes) -> InitiateTCPClientRep:
        """
        For message type 0x02, parse out:
          - client_id (str)
          - bind_address (str)
          - bind_port (uint32)
          - address_type (uint32)
          - reason (uint32)
        """
        client_id, value = MessageParser.read_string(value)
        bind_address, value = MessageParser.read_string(value)
        bind_port, value = MessageParser.read_uint32(value)
        address_type, value = MessageParser.read_uint32(value)
        reason, value = MessageParser.read_uint32(value)
        remote_addr = ''
        remote_port = 0
        if len(value) > 0:
            remote_addr, value = MessageParser.read_string(value)
            remote_port, value = MessageParser.read_uint32(value)
        return InitiateTCPClientRep(
            client_id=client_id,
            bind_address=bind_address,
            bind_port=bind_port,
            address_type=address_type,
            reason=reason,
            remote_addr=remote_addr,
            remote_port=remote_port
        )

    @staticmethod
    def parse_send_data(value: bytes) -> SendDataMessage:
        client_id, value = MessageParser.read_string(value)
        encoded_data, value = MessageParser.read_string(value)
        raw_data = base64.b64decode(encoded_data)
        return SendDataMessage(
            client_id=client_id,
            data=raw_data
        )

    @staticmethod
    def parse_initiate_bind_req(value: bytes) -> InitiateBINDReq:
        bind_id, value = MessageParser.read_string(value)
        listening_host, value = MessageParser.read_string(value)
        listening_port, value = MessageParser.read_uint32(value)
        destination_host, value = MessageParser.read_string(value)
        destination_port, value = MessageParser.read_uint32(value)
        return InitiateBINDReq(
            bind_id=bind_id,
            listening_host=listening_host,
            listening_port=listening_port,
            destination_host=destination_host,
            destination_port=destination_port
        )

    @staticmethod
    def parse_initiate_bind_rep(value: bytes) -> InitiateBINDRep:
        bind_id, value = MessageParser.read_string(value)
        listening_host, value = MessageParser.read_string(value)
        listening_port, value = MessageParser.read_uint32(value)
        reason, value = MessageParser.read_uint32(value)
        return InitiateBINDRep(
            bind_id=bind_id,
            listening_host=listening_host,
            listening_port=listening_port,
            reason=reason
        )

    @staticmethod
    def deserialize_message(encryption_key: bytes, raw_data: bytes):
        message_type, data = MessageParser.read_uint32(raw_data)
        message_length, data = MessageParser.read_uint32(data)

        payload_len = message_length - 8
        if payload_len < 0:
            raise ValueError(f"Invalid message_length {message_length}, must be at least 8")
        if len(data) < payload_len:
            raise ValueError("Not enough bytes in data for the payload")

        payload = data[:payload_len]
        leftover = data[payload_len:]

        if message_type == 0x01:
            decrypted = decrypt(encryption_key, payload)
            parsed_msg = MessageParser.parse_initiate_tcp_client_req(decrypted)
        elif message_type == 0x02:
            decrypted = decrypt(encryption_key, payload)
            parsed_msg = MessageParser.parse_initiate_tcp_client_rep(decrypted)
        elif message_type == 0x03:
            decrypted = decrypt(encryption_key, payload)
            parsed_msg = MessageParser.parse_send_data(decrypted)
        elif message_type == 0x04:
            parsed_msg = MessageParser.parse_check_in(payload)
        elif message_type == 0x05:
            decrypted = decrypt(encryption_key, payload)
            parsed_msg = MessageParser.parse_initiate_bind_req(decrypted)
        elif message_type == 0x06:
            decrypted = decrypt(encryption_key, payload)
            parsed_msg = MessageParser.parse_initiate_bind_rep(decrypted)
        elif message_type == 0x07:
            parsed_msg = CheckOutMessage()
        else:
            raise ValueError(f"Unknown message type: {hex(message_type)}")

        return leftover, parsed_msg


# --------------------------------
# 3. MessageBuilder: Creating Bytes
# --------------------------------

class MessageBuilder:
    @staticmethod
    def serialize_message(encryption_key: bytes, msg) -> bytes:
        """
        High-level build entrypoint: accept one of our named tuples and return
        the fully built+encrypted bytes (including message type, length, etc.).
        """
        value = b''
        if isinstance(msg, InitiateTCPClientReq):
            message_type = 0x01
            value = encrypt(encryption_key, MessageBuilder.build_initiate_tcp_client_req(
                msg.client_id,
                msg.destination_host,
                msg.destination_port,
                msg.listening_host,
                msg.listening_port
            ))
        elif isinstance(msg, InitiateTCPClientRep):
            message_type = 0x02
            value = encrypt(encryption_key, MessageBuilder.build_initiate_tcp_client_rep(
                msg.client_id,
                msg.bind_address,
                msg.bind_port,
                msg.address_type,
                msg.reason,
                msg.remote_addr,
                msg.remote_port
            ))
        elif isinstance(msg, SendDataMessage):
            message_type = 0x03
            value = encrypt(encryption_key, MessageBuilder.build_send_data(
                msg.client_id,
                msg.data
            ))
        elif isinstance(msg, CheckInMessage):
            message_type = 0x04
            value = MessageBuilder.build_check_in_message(
                msg.messenger_id
            )
        elif isinstance(msg, InitiateBINDReq):
            message_type = 0x05
            value = encrypt(encryption_key, MessageBuilder.build_initiate_bind_req(
                msg.bind_id,
                msg.listening_host,
                msg.listening_port,
                msg.destination_host,
                msg.destination_port
            ))
        elif isinstance(msg, InitiateBINDRep):
            message_type = 0x06
            value = encrypt(encryption_key, MessageBuilder.build_initiate_bind_rep(
                msg.bind_id,
                msg.listening_host,
                msg.listening_port,
                msg.reason
            ))
        elif isinstance(msg, CheckOutMessage):
            message_type = 0x07
            value = b''
        else:
            raise ValueError(f"Unknown message tuple type: {type(msg)}")

        return MessageBuilder.build_message(message_type, value)

    @staticmethod
    def build_message(message_type: int, value: bytes) -> bytes:
        """
        Common routine to build the 8-byte header and append encrypted payload:
          1) 4 bytes: message_type
          2) 4 bytes: total_length (header + payload)
          3) remainder: encrypt(encryption_key, plaintext_value)
        """
        message_length = 8 + len(value)
        header = struct.pack('!II', message_type, message_length)
        return header + value

    @staticmethod
    def build_string(value: str) -> bytes:
        """
        Encodes a string with a 4-byte length prefix, plus the UTF-8 data.
        """
        encoded = value.encode('utf-8')
        return struct.pack('!I', len(encoded)) + encoded

    @staticmethod
    def build_check_in_message(messenger_id: str) -> bytes:
        return MessageBuilder.build_string(messenger_id)

    @staticmethod
    def build_initiate_tcp_client_req(client_id: str,
                                      destination_host: str, destination_port: int,
                                      listening_host: str = '', listening_port: int = 0) -> bytes:
        result = (
            MessageBuilder.build_string(client_id) +
            MessageBuilder.build_string(destination_host) +
            struct.pack('!I', destination_port)
        )
        # Only a remote port forwarder appends its listening endpoint; empty
        # host means "not an RPF request" so nothing is appended.
        if listening_host:
            result += MessageBuilder.build_string(listening_host) + struct.pack('!I', listening_port)
        return result

    @staticmethod
    def build_initiate_tcp_client_rep(client_id: str,
                                      bind_address: str, bind_port: int,
                                      address_type: int, reason: int,
                                      remote_addr: str = '', remote_port: int = 0) -> bytes:
        result = (
            MessageBuilder.build_string(client_id) +
            MessageBuilder.build_string(bind_address) +
            struct.pack('!III', bind_port, address_type, reason)
        )
        if remote_addr:
            result += MessageBuilder.build_string(remote_addr) + struct.pack('!I', remote_port)
        return result

    @staticmethod
    def build_send_data(client_id: str, data: bytes) -> bytes:
        encoded_data = base64.b64encode(data).decode('utf-8')
        return (
            MessageBuilder.build_string(client_id) +
            MessageBuilder.build_string(encoded_data)
        )

    @staticmethod
    def build_initiate_bind_req(bind_id: str, listening_host: str,
                                listening_port: int, destination_host: str,
                                destination_port: int) -> bytes:
        return (
            MessageBuilder.build_string(bind_id) +
            MessageBuilder.build_string(listening_host) +
            struct.pack('!I', listening_port) +
            MessageBuilder.build_string(destination_host) +
            struct.pack('!I', destination_port)
        )

    @staticmethod
    def build_initiate_bind_rep(bind_id: str, listening_host: str,
                                listening_port: int, reason: int) -> bytes:
        return (
            MessageBuilder.build_string(bind_id) +
            MessageBuilder.build_string(listening_host) +
            struct.pack('!II', listening_port, reason)
        )

File: messenger/clients/csharp/MessengerClient/Message.cs
using System;
using System.Linq;
using System.Text;
using System.Security.Cryptography;


// ----------------------------------------------------------------------
// 1. Message Classes (now using int instead of uint)
// ----------------------------------------------------------------------
public class CheckInMessage
{
    public string MessengerId { get; }
    public CheckInMessage(string messengerId)
    {
        MessengerId = messengerId;
    }
}

public class InitiateForwarderClientReq
{
    public string ForwarderClientId { get; }
    public string IpAddress { get; }
    public int Port { get; }

    public InitiateForwarderClientReq(string forwarderClientId, string ipAddress, int port)
    {
        ForwarderClientId = forwarderClientId;
        IpAddress = ipAddress;
        Port = port;
    }
}

public class InitiateForwarderClientRep
{
    public string ForwarderClientId { get; }
    public string BindAddress { get; }
    public int BindPort { get; }
    public int AddressType { get; }
    public int Reason { get; }

    public InitiateForwarderClientRep(
        string forwarderClientId,
        string bindAddress,
        int bindPort,
        int addressType,
        int reason)
    {
        ForwarderClientId = forwarderClientId;
        BindAddress = bindAddress;
        BindPort = bindPort;
        AddressType = addressType;
        Reason = reason;
    }
}

public class SendDataMessage
{
    public string ForwarderClientId { get; }
    public byte[] Data { get; }

    public SendDataMessage(string forwarderClientId, byte[] data)
    {
        ForwarderClientId = forwarderClientId;
        Data = data;
    }
}


// ----------------------------------------------------------------------
// 2. MessageParser: Reading/Decrypting Bytes
// ----------------------------------------------------------------------
public static class MessageParser
{
    /// <summary>
    /// Read the first 4 bytes of <paramref name="data"/> as a big-endian uint32.
    /// Returns (uintValue, leftoverBytes).
    /// </summary>
    public static (uint Value, byte[] Remainder) ReadUInt32(byte[] data)
    {
        if (data.Length < 4)
            throw new ArgumentException("Not enough bytes to read a 32-bit value.");

        // Big-endian decode into a uint
        uint value = (uint)((data[0] << 24) | (data[1] << 16) | (data[2] << 8) | data[3]);
        byte[] remainder = data.Skip(4).ToArray();

        return (value, remainder);
    }

    /// <summary>
    /// Reads a length-prefixed UTF-8 string:
    ///  1) read a uint32 length
    ///  2) read 'length' bytes
    ///  3) decode UTF-8
    /// Returns (stringValue, leftoverBytes).
    /// </summary>
    public static (string Value, byte[] Remainder) ReadString(byte[] data)
    {
        var (length, remainder) = ReadUInt32(data);
        if (remainder.Length < length)
            throw new ArgumentException($"Not enough bytes to read string of length {length}.");

        string s = Encoding.UTF8.GetString(remainder, 0, (int)length);
        byte[] leftover = remainder.Skip((int)length).ToArray();

        return (s, leftover);
    }

    /// <summary>
    /// Parses the payload of a 0x04 'CheckIn' message (no decryption).
    /// </summary>
    public static CheckInMessage ParseCheckIn(byte[] value)
    {
        var (messengerId, _) = ReadString(value);
        return new CheckInMessage(messengerId);
    }

    /// <summary>
    /// Parses the payload of a 0x01 'InitiateForwarderClientReq' message.
    /// </summary>
    public static InitiateForwarderClientReq ParseInitiateForwarderClientReq(byte[] value)
    {
        var (forwarderClientId, remainder) = ReadString(value);
        var (ipAddress, remainder2) = ReadString(remainder);
        var (port, remainder3) = ReadUInt32(remainder2);

        // cast uint -> int
        return new InitiateForwarderClientReq(
            forwarderClientId,
            ipAddress,
            (int)port
        );
    }

    /// <summary>
    /// Parses the payload of a 0x02 'InitiateForwarderClientRep' message.
    /// </summary>
    public static InitiateForwarderClientRep ParseInitiateForwarderClientRep(byte[] value)
    {
        var (forwarderClientId, remainder) = ReadString(value);
        var (bindAddress, remainder2) = ReadString(remainder);
        var (bindPort, remainder3) = ReadUInt32(remainder2);
        var (addressType, remainder4) = ReadUInt32(remainder3);
        var (reason, remainder5) = ReadUInt32(remainder4);

        // cast all uint -> int
        return new InitiateForwarderClientRep(
            forwarderClientId,
            bindAddress,
            (int)bindPort,
            (int)addressType,
            (int)reason
        );
    }

    /// <summary>
    /// Parses the payload of a 0x03 'SendDataMessage'.
    /// </summary>
    public static SendDataMessage ParseSendData(byte[] value)
    {
        var (forwarderClientId, remainder) = ReadString(value);
        var (encodedData, remainder2) = ReadString(remainder);

        byte[] rawData = Convert.FromBase64String(encodedData);
        return new SendDataMessage(
            forwarderClientId,
            rawData
        );
    }

    /// <summary>
    /// High-level parse entrypoint:
    ///   1) read the message_type (uint32)
    ///   2) read the message_length (uint32)
    ///   3) slice out the encrypted payload
    ///   4) decrypt or parse plaintext
    /// Returns (leftoverBytes, parsedMessage).
    /// </summary>
    public static (byte[] leftover, object parsedMessage) DeserializeMessage(
        byte[] encryptionKey,
        byte[] rawData)
    {
        // 1) Read the message type
        var (messageType, dataAfterType) = ReadUInt32(rawData);

        // 2) Read the total message length
        var (messageLength, dataAfterLength) = ReadUInt32(dataAfterType);

        // 3) Extract the (messageLength - 8) payload bytes
        int payloadLen = (int)messageLength - 8;
        if (dataAfterLength.Length < payloadLen)
            throw new ArgumentException("Not enough bytes in data for the payload.");

        byte[] payload = dataAfterLength.Take(payloadLen).ToArray();
        byte[] leftover = dataAfterLength.Skip(payloadLen).ToArray();

        // 4) Decrypt or parse plaintext, depending on message type
        object parsedMsg;
        switch (messageType)
        {
            case 0x01:
                {
                    byte[] decrypted = MessengerClient.Crypto.Decrypt(encryptionKey, payload);
                    parsedMsg = ParseInitiateForwarderClientReq(decrypted);
                    break;
                }
            case 0x02:
                {
                    byte[] decrypted = MessengerClient.Crypto.Decrypt(encryptionKey, payload);
                    parsedMsg = ParseInitiateForwarderClientRep(decrypted);
                    break;
                }
            case 0x03:
                {
                    byte[] decrypted = MessengerClient.Crypto.Decrypt(encryptionKey, payload);
                    parsedMsg = ParseSendData(decrypted);
                    break;
                }
            case 0x04:
                {
                    // According to the Python version, we do NOT decrypt for 0x04
                    parsedMsg = ParseCheckIn(payload);
                    break;
                }
            default:
                throw new ArgumentException($"Unknown message type: 0x{messageType:X}");
        }

        return (leftover, parsedMsg);
    }
}


// ----------------------------------------------------------------------
// 3. MessageBuilder: Creating/Encrypting Bytes
// ----------------------------------------------------------------------
public static class MessageBuilder
{
    /// <summary>
    /// Given a message object (one of our 4 types), build the fully formed
    /// byte array (header + possibly encrypted payload).
    /// </summary>
    public static byte[] SerializeMessage(byte[] encryptionKey, object msg)
    {
        byte[] payload;
        uint messageType;

        switch (msg)
        {
            case InitiateForwarderClientReq req:
                messageType = 0x01;
                payload = MessengerClient.Crypto.Encrypt(
                    encryptionKey,
                    BuildInitiateForwarderClientReq(
                        req.ForwarderClientId,
                        req.IpAddress,
                        req.Port
                    )
                );
                break;

            case InitiateForwarderClientRep rep:
                messageType = 0x02;
                payload = MessengerClient.Crypto.Encrypt(
                    encryptionKey,
                    BuildInitiateForwarderClientRep(
                        rep.ForwarderClientId,
                        rep.BindAddress,
                        rep.BindPort,
                        rep.AddressType,
                        rep.Reason
                    )
                );
                break;

            case SendDataMessage sdm:
                messageType = 0x03;
                payload = MessengerClient.Crypto.Encrypt(
                    encryptionKey,
                    BuildSendData(
                        sdm.ForwarderClientId,
                        sdm.Data
                    )
                );
                break;

            case CheckInMessage cim:
                messageType = 0x04;
                // The Python code does not encrypt CheckInMessage
                payload = BuildCheckInMessage(cim.MessengerId);
                break;

            default:
                throw new ArgumentException($"Unknown message type: {msg.GetType().Name}");
        }

        return BuildMessage(messageType, payload);
    }

    /// <summary>
    /// Packs [message_type, total_length, payload] together in big-endian format.
    /// </summary>
    public static byte[] BuildMessage(uint messageType, byte[] payload)
    {
        // total_length = 8 (header) + payload length
        uint messageLength = (uint)(8 + payload.Length);

        // Build the 8-byte header (all big-endian)
        byte[] header = new byte[8];
        WriteUInt32(header, 0, messageType);    // message_type
        WriteUInt32(header, 4, messageLength);  // total_length

        // Concatenate header + payload
        return Combine(header, payload);
    }

    /// <summary>
    /// Encodes a string with a 4-byte length prefix (big-endian), plus UTF-8 data.
    /// </summary>
    public static byte[] BuildString(string value)
    {
        byte[] encoded = Encoding.UTF8.GetBytes(value);
        byte[] lengthBytes = new byte[4];
        WriteUInt32(lengthBytes, 0, (uint)encoded.Length);

        return Combine(lengthBytes, encoded);
    }

    public static byte[] BuildCheckInMessage(string messengerId)
    {
        return BuildString(messengerId);
    }

    public static byte[] BuildInitiateForwarderClientReq(
        string forwarderClientId,
        string ipAddress,
        int port)
    {
        var part1 = BuildString(forwarderClientId);
        var part2 = BuildString(ipAddress);

        // We'll still write it as a uint on the wire
        byte[] part3 = new byte[4];
        WriteUInt32(part3, 0, (uint)port);

        return Combine(part1, part2, part3);
    }

    public static byte[] BuildInitiateForwarderClientRep(
        string forwarderClientId,
        string bindAddress,
        int bindPort,
        int addressType,
        int reason)
    {
        var part1 = BuildString(forwarderClientId);
        var part2 = BuildString(bindAddress);

        // Next 3 fields => 3*4 bytes => 12 bytes
        byte[] part3 = new byte[12];
        WriteUInt32(part3, 0, (uint)bindPort);
        WriteUInt32(part3, 4, (uint)addressType);
        WriteUInt32(part3, 8, (uint)reason);

        return Combine(part1, part2, part3);
    }

    public static byte[] BuildSendData(
        string forwarderClientId,
        byte[] data)
    {
        var part1 = BuildString(forwarderClientId);
        string encodedData = Convert.ToBase64String(data);
        var part2 = BuildString(encodedData);

        return Combine(part1, part2);
    }

    // ----------------------------------------------------------------------
    // Helper: Write a uint32 to a byte array in big-endian
    // ----------------------------------------------------------------------
    public static void WriteUInt32(byte[] buffer, int offset, uint value)
    {
        buffer[offset] = (byte)((value >> 24) & 0xFF);
        buffer[offset + 1] = (byte)((value >> 16) & 0xFF);
        buffer[offset + 2] = (byte)((value >> 8) & 0xFF);
        buffer[offset + 3] = (byte)(value & 0xFF);
    }

    // ----------------------------------------------------------------------
    // Helper: Concatenate multiple byte arrays
    // ----------------------------------------------------------------------
    public static byte[] Combine(params byte[][] arrays)
    {
        int totalLength = 0;
        foreach (var arr in arrays)
            totalLength += arr.Length;

        byte[] result = new byte[totalLength];
        int offset = 0;

        foreach (var arr in arrays)
        {
            Buffer.BlockCopy(arr, 0, result, offset, arr.Length);
            offset += arr.Length;
        }

        return result;
    }
}
